src/tcache.c

```diff
@@ -390,40 +426,68 @@ tcache_event(tsd_t *tsd) {
 	}
 
 	tcache_slow_t *tcache_slow = tsd_tcache_slowp_get(tsd);
-	szind_t szind = tcache_slow->next_gc_bin;
-	bool is_small = (szind < SC_NBINS);
-	cache_bin_t *cache_bin = &tcache->bins[szind];
+	assert(tcache_slow != NULL);
 
-	if (tcache_bin_disabled(szind, cache_bin, tcache_slow)) {
-		goto label_done;
+	/* When the new tcache gc is not enabled, GC one bin at a time. */
+	if (!opt_experimental_tcache_gc) {
+		szind_t szind = tcache_slow->next_gc_bin;
+		bool is_small = (szind < SC_NBINS);
+		tcache_gc(tsd, tcache_slow, tcache, szind, is_small);
+		tcache_slow->next_gc_bin++;
+		if (tcache_slow->next_gc_bin == tcache_nbins_get(tcache_slow)) {
+			tcache_slow->next_gc_bin = 0;
+		}
+		return;
 	}
 
-	tcache_bin_flush_stashed(tsd, tcache, cache_bin, szind, is_small);
-	cache_bin_sz_t low_water = cache_bin_low_water_get(cache_bin);
-	if (low_water > 0) {
-		if (is_small) {
-			tcache_gc_small(tsd, tcache_slow, tcache, szind);
-		} else {
-			tcache_gc_large(tsd, tcache_slow, tcache, szind);
+	nstime_t now;
```

Comment:
Slightly different steps that's simpler:
```
nstime_t now;
nstime_copy(&now, &tcache_slow->last_gc_time);
nstime_update(&now);
```
Then This way you won't have to handle non monotonic clock yourself.

Also use these facilities to read the time instead of direct access, for additional sanity checks:
https://github.com/jemalloc/jemalloc/blob/8c54637f8c7a98bbaec6ee38229a904bbf22170c/src/nstime.c#L60